mantidproject/mantid
# Pearl Powder Diffraction Script - Prep work

Description of work.
As part of #17831 - Migrating Powder Diffraction instruments to use a common script.
This PR is to merge in a working implementation with old API support before starting the migration of additional instruments into the script. 

I am open to any suggestions or changes - however as part of the ongoing process I'll add them into the next PR instead of modifying this PR (unless they affect the data being generated) as this script is in a tested working state. 

This PR lays down the prep work for the next step, migration of POLARIS. It does the following:
- Removed all globals (except for a global debug flag and a small work around for the ADS overwriting similar names)
- Made there a single entry/exit point for various public calls
- Removed the requirement that certain APIs be called in an (undocumented) order to setup the directories
- Separated public and private API using notation 
- Implemented an Abstract instrument and PEARL specific overrides and hooks which helps contain instrument specific workarounds or parameters
- Completely moved and refactored the existing code out of `pearl_routines`
- Migrated everything to Python 2/3 (I may have missed one please let me know if I have)
- Rewrote `pearl_routines` as a wrapper around the new implementation and verified compatibility with existing tests
- Verified new data matched old data, then rewrote system tests to correctly use reference workspaces with the old API
- Implemented new unit tests for functions which are unit testable in both `AbstractInstrument` and `PEARL` instrument

---

There are still some steps still to do as part of future issues:
- Generate absorption corrections does not work as the existing implementation is incorrect and there are no unit tests to verify the behaviour. This will be fixed by working through the history of existing correction files.
- A user prefs file will be created and read for each instrument. This means that users can edit some of the advanced settings or add common settings (i.e. for their machine) in a separate script. This discourages modification to the current scripts. 
- Creating a silicon calibration has not data to verify the new API against so will need additional testing. 
- Using system tests directly calling the new API - there should be no difference in the algorithms only the entry point hence this isn't implemented yet. 
- Documentation - this is because the new API is still in flux especially when new instruments need to be added and the entry points might change

**To test:**
System tests should pass against the reference data

Run the following:

``` python
import PearlPowder_Pearl

calibration_dir =  *to_set*
raw_data_dir =  *to_set*
output_dir = *to_set*
tt_mode = "TT70"

pearl_obj = PearlPowder_Pearl.Pearl(user_name="Test", calibration_dir=calibration_dir,
                                    raw_data_dir=raw_data_dir, output_dir=output_dir, tt_mode=tt_mode)
output = pearl_obj.focus(run_number=92476, focus_mode="mods")
```

I recommend setting `calibration_dir` and `raw_data_dir` to 
`*SystemTestData*/PEARL/Focus_Test/Calibration/` and 
`*SystemTestData*/PEARL/Focus_Test/RawFiles/` respectively 
Also don't forget to set the output directory to your choice.

The paths **must** end with a `/` or `\\` character (remembering that `\` escapes so Windows paths need to be separated by `\\`

Fixes #17839 .



---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
